public/src/main/java/com/revenuecat/purchases/models/StoreProduct.kt

```diff
@@ -137,7 +135,29 @@ data class StoreProduct(
      * For Amazon, the original Product can be obtained using `StoreProduct.amazonProduct`
      */
     val originalJson: JSONObject
-) : Parcelable {
+}
+
+@Parcelize
+@TypeParceler<JSONObject, JSONObjectParceler>()
+data class StoreProductImpl(
+    override val sku: String,
+    override val type: ProductType,
+    override val price: String,
+    override val priceAmountMicros: Long,
+    override val priceCurrencyCode: String,
+    override val originalPrice: String?,
+    override val originalPriceAmountMicros: Long,
+    override val title: String,
+    override val description: String,
+    override val subscriptionPeriod: String?,
+    override val freeTrialPeriod: String?,
+    override val introductoryPrice: String?,
+    override val introductoryPriceAmountMicros: Long,
+    override val introductoryPricePeriod: String?,
+    override val introductoryPriceCycles: Int,
+    override val iconUrl: String,
+    override val originalJson: JSONObject
+) : StoreProduct, Parcelable {
```

Comment:
Thanks for reviewing! I'll answer here to both of your comments to keep it in a thread:
The main idea wasn't to make it available to developers but rather use this internally (although I think we can surely expose the information to devs to).

Basically my idea was that we'll decide internally whether we can use the new BC5 interfaces for purchases or not.
This means that for kicking of a purchase flow we need a few more pieces of data ProductDetails, and offerToken).
* Deciding with [params to use for purchaseFlow](https://github.com/RevenueCat/purchases-android/blob/04d755311c1dd463279ad63f95303b8cda06f84b/feature/google/src/bc5SupportDev/java/com/revenuecat/purchases/google/BillingWrapper.kt#L328-L333)
* How to map StoreProducts to packages in getOfferings
* Decide whether we need to send [pricingPhases](https://github.com/RevenueCat/purchases-android/blob/04d755311c1dd463279ad63f95303b8cda06f84b/common/src/main/java/com/revenuecat/purchases/common/Backend.kt#L183) to the backend

I'm not too familiar with extensions, but I understand the idea would be to add all the BC5 related fields inside the google feature so that they're only visible inside this compilation unit. We'd basically take all these decisions based whether the new BC5 fields are null or not, right?
My only concern (which I also have with my current approach) is that the backend needs some extra information on the StoreProduct to know whether it was a BC5 purchase or not, which makes us need to pull some of this out of just the google feature.

Reply:
thinking through the code more, i am feeling like we might want to get the purchase flow working before we know how we want this.

i started a little work making `useBC5` a parameter on the purchase configuration that would remove the need to use `StoreProduct` to branch between BC5 flows.

if we were to go with that, i think we still have a question about what to do with the `StoreProduct` public API and whether we can squeeze BC5 things into it or not...but that's what i think we will discover when we finish up the purchase flow work.

Reply:
I looked up the way Kotlin extensions work a bit more and I honestly don't fully understand what the reason was to prefer this over using interfaces and subclasses.
I understand that downcasting might be a bit of a pain - but right now it's relying on a generic map to store data and every time the property is accessed, converting data from that JSON object.
That doesn't really seem neither efficient, nor very safe.

In the end exposing the interface in the public and then using different implementation classes would have pretty much the same effect, but being more efficient because we can actually back the the added getters with a normal field implementation (and we also don't have to expose the JSONObject in the public interface).
True that downcasting might prove a bit more painful, but that should only be required in the direct implementations of shared interfaces (e.g. BillingAbstract main methods).

Re: the useBC5 flag on the client side means that we're putting more of a burden on the developer - they have to consciously enable a parameter with no apparent gain - nothing has changed in the API for them and there are no new features. So why can't we control this for them?

Reply:
If I remember correctly, one of the main reasons why we did it that way was to make the migration to the new major version of the SDK easier.

We were exposing the `SKUDetails` directly, so if someone were accessing `skuDetails.introPricePeriod` in v2, they would have to change it to `(product as GoogleStoreProductImpl).skuDetails.introPricePeriod` in order to access `introPricePeriod`, which is only available in Google. We were planning on making `StoreProduct` to only contain what was common in both Google and Amazon, so `StoreProduct` wouldn't have an introPricePeriod since Amazon doesn't have it.

So we ended up taking the approach of making a bunch of things nullable (for Amazon only), and not having an `AmazonImpl` and a `GoogleStoreProductImpl`, just to make sure the Google implementation didn't require a lot of downcasting.

I am not married to this approach and we could revisit it if we release this change as a major. So far, I believe this PR I think only has an effect on the public API in case someone is instantiating `StoreProduct` directly, so it would need to be released as a major too.

Reply:
Ok, I'm not sure why we'd want to expose SKUDetails though, having devs rely on this is pretty bad if we want to switch over to BC5 where this data won't be available anymore.
Do we know specifically what people use this for? What data isn't available on StoreProduct directly? 
I totally agree with the approach of making StoreProduct only have all the common attributes, and I think we should deprecate it here and then in a major version move it down to the BC4StoreProduct instead.
Hopefully we can make all data available for people on StoreProduct itself based on both BC4 and BC5 underlying classes.
E.g.: Our idea was to unify all the different prices in a pricingPhases list like we have in BC5. This should cover all these different prices. We can still provide the old fields as backwards compatibility although they should be deprecated, too, given that we can't really cover all use cases with this (pricingPhases might have more than 3 entries).

Do you think that would be doable - any concerns with this approach?

Re: instantiating StoreProduct directly - to me that seems like a misuse of the SDK, but I'm not fully aware of all the use cases. Is there any reason why someone would want to do this?

Reply:
> Ok, I'm not sure why we'd want to expose SKUDetails though, having devs rely on this is pretty bad if we want to switch over to BC5 where this data won't be available anymore.

That was just for flexibility, in case someone was migrating from their own implementation to RC, to make it easier. Or upgrading from a previous version of this SDK, where our public API relied directly on SKUDetails. There are some fields in the Amazon product that are not exposed though StoreProduct though, so it's more useful in that case.

> Re: instantiating StoreProduct directly - to me that seems like a misuse of the SDK, but I'm not fully aware of all the use cases. Is there any reason why someone would want to do this?

Unit testing is the first thing that comes to mind. Someone might be instantiating directly StoreProduct when testing.

> E.g.: Our idea was to unify all the different prices in a pricingPhases list like we have in BC5. This should cover all these different prices. We can still provide the old fields as backwards compatibility although they should be deprecated, too, given that we can't really cover all use cases with this (pricingPhases might have more than 3 entries).
So for now there would be a new `pricingPhases` list in deprecated StoreProduct (which will be nullable for non BC5). And then have a `BC4StoreProduct`, `BC5StoreProduct` and an `AmazonStoreProduct`? Then a `StoreProduct` that has only the common stuff in the three of them. Correct?